core/string/ustring.cpp

```diff
@@ -531,10 +531,16 @@ String &String::operator+=(const String &p_str) {
 
 	resize(lhs_len + rhs_len + 1);
 
-	const char32_t *src = p_str.get_data();
+	const char32_t *src = p_str.ptr();
 	char32_t *dst = ptrw() + lhs_len;
 
-	memcpy(dst, src, (rhs_len + 1) * sizeof(char32_t));
+	if (ptr() == src) { // Adding the string to itself.
+		// Copy the terminating null separately as it overlaps the destination (`memcpy` has undefined behavior for overlapping blocks).
+		memcpy(dst + rhs_len, src + lhs_len, sizeof(char32_t));
```

Comment:
I guess we can keep only `memcpy(dst, src, rhs_len * sizeof(char32_t));` and directly set last char to `0` in all cases, without extra `if`.